### DISCUSSION TITLE
Interrupts in a thread ?

### DISCUSSION THREAD

**Discussion by Gordon999** [micropython/micropython#9392] [Using MicroPython]
I tried the following script but the interrupts won't work, they are fine outside the thread and I can read the value of the GP inputs in the thread BUT the interrupts won't work in the thread.

The` script generates 2 pwm signals and should count them on other inputs

Any thoughts ?

Using raspberry pico.
```
import time
from machine import I2C, PWM, freq,Pin
import _thread

# start pwm outputs
fwd_pwm  = PWM(Pin(26)) # connected to GP10
fwd_pwm.freq(50)        # 50Hz
fwd_pwm.duty_u16(32000) # 50% duty cycle
bac_pwm  = PWM(Pin(27)) # connected to GP12
bac_pwm.freq(50)        # 50Hz
bac_pwm.duty_u16(15000) # 25% duty cycle

#initialise variables
start = time.ticks_us()
count1 = 0
count2 = 0

# thread
def count():
    global count1,count2
    def counter1(pin):
        global count1
        if pinA2.value() == 1 :
            count1 += 1
        else:
            count1 -= 1
    def counter2(pin):
        global count2
        if pinA1.value() == 1 :
            count2 += 1
        else:
            count2 -= 1
    # setup interrupts
    pinA1 = Pin(10, Pin.IN)
    pinA1.irq(trigger=Pin.IRQ_FALLING, handler=counter1)
    pinA2 = Pin(12, Pin.IN)
    pinA2.irq(trigger=Pin.IRQ_FALLING, handler=counter2)
       
        
_thread.start_new_thread(count,())

while True:
    # display counters
    time.sleep(1)
    if time.ticks_us() - start > 1000000: # display counts every second
        print(count1,count2)
        start = time.ticks_us()```

**Comment by robert-hh** [comment]
Please use triple backticks ``` to enclose code sections

**Reply by Gordon999** [reply]
Thanks

**Comment by jimmo** [comment]
This sounds like it might be the same as https://github.com/micropython/micropython/issues/9124

**Reply by Gordon999** [reply]
Thankyou. It looks like the same issue.
I tried modifying my script and adding hard=True, and using your micropython script but it didn't solve it.
Do l need to modify machine_pin.c to resolve it ?

**Reply by jimmo** [reply]
It's a bigger change than just machine_pin.c

**Reply by Gordon999** [reply]
Thanks for the response.
So what happens next ? Will it get fixed ?

**Reply by jimmo** [reply]
It's on the TODO list, it's definitely something that needs to be fixed. Not likely to happen for v1.20 though. Contributions would definitely be welcome!

**Comment by ctimmer** [comment]
Just a comment on a problem unrelated to the interrupt issue. Using standard math operations on ticks_us/ticks_ms caused me hours of debug time:
```
if time.ticks_us() - start > 1000000: # display counts every second
```
ticks_diff should be used. Here is the explanation:
[ticks_ms](https://docs.micropython.org/en/latest/library/time.html#time.ticks_ms)

Curt
